src/main/java/org/opensearch/neuralsearch/query/HybridQueryDocIdStream.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.opensearch.neuralsearch.query;
+
+import lombok.RequiredArgsConstructor;
+import lombok.Setter;
+import org.apache.lucene.search.CheckedIntConsumer;
+import org.apache.lucene.search.DocIdStream;
+import org.apache.lucene.util.FixedBitSet;
+
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * This class is used to create a DocIdStream for HybridQuery
+ */
+@RequiredArgsConstructor
+public class HybridQueryDocIdStream extends DocIdStream {
+    private final HybridBulkScorer hybridBulkScorer;
+    @Setter
+    private int base;
+
+    /**
+     * Iterate over all doc ids and collect each doc id with leaf collector
+     * @param consumer consumer that is called for each accepted doc id
+     * @throws IOException in case of IO exception
+     */
+    @Override
+    public void forEach(CheckedIntConsumer<IOException> consumer) throws IOException {
+        // bitset that represents matching documents, bit is set (1) if doc id is a match
+        FixedBitSet matchingBitSet = hybridBulkScorer.getMatching();
+        long[] bitArray = matchingBitSet.getBits();
+        // iterate through each block of 64 documents (since each long contains 64 bits)
+        for (int idx = 0; idx < bitArray.length; idx++) {
+            long bits = bitArray[idx];
+            while (bits != 0L) {
+                // find position of the rightmost set bit (1)
+                int numberOfTrailingZeros = Long.numberOfTrailingZeros(bits);
+                // calculate actual document ID within the window
+                // idx << 6 is equivalent to idx * 64 (block offset)
+                // numberOfTrailingZeros gives position within the block
+                final int docIndexInWindow = (idx << 6) | numberOfTrailingZeros;
+                float[][] windowScores = hybridBulkScorer.getWindowScores();
+                for (int subQueryIndex = 0; subQueryIndex < windowScores.length; subQueryIndex++) {
+                    if (Objects.isNull(windowScores[subQueryIndex])) {
+                        continue;
+                    }
+                    float scoreOfDocIdForSubQuery = windowScores[subQueryIndex][docIndexInWindow];
```

Comment:
While there's a null check for the array `windowScores[subQueryIndex])` itself, but there's no check for individual elements within it before accessing `windowScores[subQueryIndex][docIndexInWindow]`?

Reply:
there is a valid scenario when `windowScores[subQueryIndex]` can be null, if sub-query has no matches for one window. As this is array of primitive `float` individual elements will be initialized with 0.0 values. Maybe I misunderstood the question.

Reply:
Thanks for the clarification!